compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -733,6 +859,33 @@ abstract class GenJSCode extends plugins.PluginComponent
           newBody)(methodDef.optimizerHints, None)(methodDef.pos)
     }
 
+    /** Moves all statements after the super constructor call.
+     *
+     *  This is used for the primary constructor of a Scala.js-defined JS
+     *  class, because those cannot access `this` before the super constructor
+     *  call.
+     *
+     *  scalac inserts statements before the super constructor call for early
+     *  initializers and param accessor initializers (including val's and var's
+     *  declared in the params). We move those after the super constructor
+     *  call, and are therefore executed later than for a Scala class.
+     */
+    private def moveAllStatementsAfterSuperConstructorCall(
+        body: js.Tree): js.Tree = {
+      val bodyStats = body match {
+        case js.Block(stats) => stats
+        case _               => body :: Nil
+      }
+
+      val (beforeSuper, superCall :: afterSuper) =
+        bodyStats.span(!_.isInstanceOf[js.JSSuperConstructorCall])
+
+      js.Block(
+          superCall ::
+          beforeSuper :::
+          afterSuper)(body.pos)
+    }
+
```

Comment:
As if we ever turn off asserts ^^

I initially wanted to do something like that, but stuff like `$outer` initialization is not very pretty. In general, anyway, I think we should move anything that scalac throws at us. It doesn't seem useful to prevent anything, actually. Or maybe just refuse to move `VarDef`s, since in theory they could declare a local value that would be used in a parameter to the super constructor (although I don't see how Scala source code could result in such a pattern).